storage/storage.go

```diff
@@ -0,0 +1,124 @@
+package storage
+
+import (
+	"github.com/dgraph-io/badger"
+	"github.com/pkg/errors"
+
+	"github.com/iorhachovyevhen/dsss/models"
+)
+
+var (
+	ErrIDNotFound    = errors.New("ID not found")
+	ErrAlreadyUsedID = errors.New("ID already is used")
+	ErrInvalidID     = errors.New("Invalid ID")
+)
+
+type DataKeeper interface {
+	Add(data models.Data) (models.ID, error)
+	Read(key []byte) (models.Data, error)
+	Delete(key []byte) error
+
+	Close() error
+}
+
+type Storage struct {
+	db *badger.DB
+}
+
+func NewOptions() badger.Options {
+	return badger.DefaultOptions("")
+}
+
+func NewStorageWithOptions(opts badger.Options) *Storage {
+	return &Storage{
+		db: openDB(opts),
+	}
+}
+
+func NewDefaultStorage(savePath string) *Storage {
+	return &Storage{
+		db: openDB(badger.DefaultOptions(savePath)),
+	}
+}
+
+func openDB(opt badger.Options) *badger.DB {
+	db, err := badger.Open(opt)
+	if err != nil {
+		panic(err)
+	}
+
+	return db
+}
+
+func (s *Storage) Add(data models.Data) (models.ID, error) {
+	_, err := s.Read(data.ID())
+	if err == nil {
+		return nil, ErrAlreadyUsedID
+	}
+
+	err = s.db.Update(func(txn *badger.Txn) error {
+		val, err := data.MarshalBinary()
+		if err != nil {
+			return err
+		}
+
+		return txn.Set(data.ID(), val)
+	})
+
+	if err != nil {
+		return nil, errors.New("writing data finished with error: " + err.Error())
+	}
+
+	return data.ID(), nil
+}
+
+func (s *Storage) Read(key []byte) (models.Data, error) {
+	dt, err := models.DataTypeFromID(key)
+	if err != nil {
+		return nil, err
+	}
+
+	data := models.NewEmptyData(dt)
+
+	err = s.db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get(key)
+		if err != nil {
+			return err
+		}
+
+		err = item.Value(func(val []byte) error {
+			return data.UnmarshalBinary(val)
+		})
+
+		return err
+	})
+
+	if err != nil {
+		if err == badger.ErrKeyNotFound {
+			return nil, ErrIDNotFound
+		}
+		return nil, errors.New("reading finished with error: " + err.Error())
+	}
+
+	return data, nil
+}
+
+func (s *Storage) Delete(key []byte) error {
+	_, err := s.Read(key)
```

Comment:
This method will UnmarshalBinary. So you will load full data to check that key is not exist. You need to use something like 'contains' or 'exists' method. If badger doesn't allow to do that, so let's only try to delete, I think badger will return error by self.

Reply:
if I will try to delete to check if exist an item then it will be a one-off check :D

Reply:
Are u sure that "item" doesn't" load all data?